fix: adding a -- separator for spi options

shawkins opened the pull request:
closes: #39063

<!---
Please read https://github.com/keycloak/keycloak/blob/main/CONTRIBUTING.md and follow these guidelines when contributing to Keycloak
-->

shawkins committed:
fix: adding a -- separator for spi options

closes: #39063

Signed-off-by: Steve Hawkins <shawkins@redhat.com>

vmuzikar reviewed (commented):
> are you ok with this being a breaking change from the perspective of users who are relying on auto-builds with -provider, -provider-default, -enabled? If so I can add a note for that.

Yes, please add a note to the upgrading guide about that. Maybe with some reasoning about it.

> And / or do you want a warning for now for the possibly ambiguous cases?

You mean to the docs? Definitely yes. I know you've done some updates to the docs but IMHO we need to be much more explicit. Basically to explain the difference between the two forms and the auto build behaviour. It might very easily get confusing to the users.

Additionally, I think we should update all docs occurences to the `--` format.

shawkins committed:
updating docs to the new format

Signed-off-by: Steve Hawkins <shawkins@redhat.com>
# Conflicts:
#	docs/guides/high-availability/examples/generated/keycloak-ispn.yaml
#	docs/guides/high-availability/examples/generated/keycloak.yaml

shawkins committed:
adding a warning for ambiguous spi options

also adding a note about the change

Signed-off-by: Steve Hawkins <shawkins@redhat.com>
# Conflicts:
#	docs/documentation/upgrading/topics/changes/changes-26_3_0.adoc

shawkins commented:
Updated with a warning when changed the docs to use the new format.

There's still a reference to spi-authentication-sessions-map-auth-sessions-limit (now spi-authentication-sessions--map--auth-sessions-limit) that should be removed correct?

shawkins committed:
internally using the new spi options

also adding a deprecation notice

Signed-off-by: Steve Hawkins <shawkins@redhat.com>

mabartos commented on docs/documentation/upgrading/topics/changes/changes-26_3_0.adoc:
```diff
@@ -80,6 +84,15 @@ the user account will be updated to (un)mark the email as verified.
+To resolve this ambiguity, and any potential ambiguity involving SPI and provider names, a new SPI option format was introduced where the scopes and suffix are separated by -- instead of -. 
+
+An SPI property ending in -enabled, -provider-default, or -provider will only be treated as buildtime if it is using the new format, e.g. `spi-<spi-name>--<provider-name>--enabled` will be recognized as a buildtime option. 
+Options using the legacy format and ending in ending in -enabled, -provider-default, or -provider will instead be treated as runtime and a warning will be emitted.   
```
```suggestion
Options using the legacy format and ending in `-enabled`, `-provider-default`, or `-provider` will instead be treated as runtime and a warning will be emitted.   
```

mabartos reviewed (commented):
@shawkins Nice work! Using `--` separator makes these SPI separations more visible and I like it more than before. 

Just added some small comments that might improve things a little bit more IMO.

shawkins committed:
Apply suggestions from code review

Co-authored-by: Martin Bartoš <mabartos@redhat.com>
Signed-off-by: Steven Hawkins <shawkins@redhat.com>

shawkins commented on core/src/main/java/org/keycloak/Config.java:
```diff
@@ -235,6 +235,10 @@ public static interface Scope {
         Scope scope(String... scope);
 
+        /**
+         * @deprecated to be removed
```
Updated, and also we can't use the annotation properties in core because we're forcing Java 8 compatibility.

mabartos reviewed (commented):
@shawkins Looks good, but one last thing - we should move these upgrading notes you provided to the "Notable changes" section and probably not to the "Breaking changes". Or is there anything I missed?

mabartos commented on core/src/main/java/org/keycloak/Config.java:
```diff
@@ -235,6 +235,13 @@ public static interface Scope {
         Scope scope(String... scope);
 
+        /**
+         * @deprecated since 26.2.3, to be removed
```
```suggestion
         * @deprecated since 26.2.6, to be removed
```

nit: I believe it'll fit into the 26.2.6

mabartos reviewed (dismissed):
@shawkins Ok, LGTM, thanks!

shawkins committed:
correcting options output

adding + + inlining where needed

Signed-off-by: Steve Hawkins <shawkins@redhat.com>

mabartos reviewed (dismissed):
LGTM

vmuzikar reviewed (commented):
@shawkins Sorry for the late review and thank you for the changes!

Overall LGTM but might be good to add some tests for checking the selected SPI options are correctly treated as buildtime.
Additionally, I wonder if we should have a test for the env var form, if `__` works as expected.

shawkins commented:
> Overall LGTM but might be good to add some tests for checking the selected SPI options are correctly treated as buildtime.

We already had a test or two along those lines (see StartCommandDistTest.errorSpiBuildtimeAtRuntime for example), and the arquilian logic relies on implicit rebuilds due to provider option changes.

> Additionally, I wonder if we should have a test for the env var form, if __ works as expected.

I can certainly add something. The replacement logic in KcEnvConfigSource is now quite straight-forward all _ in a non-wildcard key will become -, so there's nothing special about this case.

shawkins committed:
Merge branch 'main' into iss39063

Signed-off-by: Steven Hawkins <shawkins@redhat.com>

shawkins committed:
adding test showing the env mapping with __

Signed-off-by: Steve Hawkins <shawkins@redhat.com>

vmuzikar commented:
> Quarkus? These entries are only handled by Keycloak logic.

Yeah, but I meant some too smart interceptors maybe. Still I'd feel safer with some tests. :)

vmuzikar reviewed (dismissed):
LGTM, thank you.

shawkins committed:
Merge remote-tracking branch 'upstream' into iss39063

shawkins commented:
@mabartos @vmuzikar the rebase was trivial, just wanted to check if either of you want to review again.

mabartos reviewed (approved):
LGTM

pl-me commented:
@shawkins Hi, I noticed you have changed MicroProfileScope.getPropertyNames() to filter keys using double-dash-separator only. That means it is a breaking change if you rely on getPropertyNames() and still using single-dash options. Am I right?
As getPropertyNames() is now deprecated I guess I should move away from using that function anyway. I just wanted to make sure your change was intentional.
